File: cp/java/10_only_number.java
import java.io.*;
import java.math.BigInteger;
import java.util.Scanner;

public class RSA {
    
    // Function to calculate gcd
    public static BigInteger gcd(BigInteger a, BigInteger b) {
        if (b.equals(BigInteger.ZERO)) {
            return a;
        }
        return gcd(b, a.mod(b));
    }

    // Function for modular exponentiation (base^exp % mod)
    public static BigInteger modPow(BigInteger base, BigInteger exp, BigInteger mod) {
        BigInteger result = BigInteger.ONE;
        while (exp.compareTo(BigInteger.ZERO) > 0) {
            if (exp.mod(BigInteger.TWO).equals(BigInteger.ONE)) {
                result = result.multiply(base).mod(mod);
            }
            base = base.multiply(base).mod(mod);
            exp = exp.divide(BigInteger.TWO);
        }
        return result;
    }

    // Function to find modular inverse using Extended Euclidean Algorithm
    public static BigInteger modInverse(BigInteger e, BigInteger phi) {
        BigInteger[] result = extendedGCD(e, phi);
        BigInteger t = result[1]; // t is the modular inverse
        if (t.compareTo(BigInteger.ZERO) < 0) {
            t = t.add(phi);
        }
        return t;
    }

    // Extended Euclidean Algorithm
    public static BigInteger[] extendedGCD(BigInteger a, BigInteger b) {
        if (b.equals(BigInteger.ZERO)) {
            return new BigInteger[] { a, BigInteger.ONE, BigInteger.ZERO };
        }
        BigInteger[] result = extendedGCD(b, a.mod(b));
        BigInteger gcd = result[0];
        BigInteger x = result[2];
        BigInteger y = result[1].subtract(a.divide(b).multiply(result[2]));
        return new BigInteger[] { gcd, x, y };
    }

    public static void main(String[] args) throws FileNotFoundException {
        Scanner input = new Scanner(new File("input.txt"));
        PrintWriter output = new PrintWriter(new File("output.txt"));

        // Read primes p and q from file
        BigInteger p = input.nextBigInteger();
        BigInteger q = input.nextBigInteger();
        
        // Calculate n = p * q and phi = (p-1)*(q-1)
        BigInteger n = p.multiply(q);
        BigInteger phi = (p.subtract(BigInteger.ONE)).multiply(q.subtract(BigInteger.ONE));

        // Choose e such that 1 < e < phi and gcd(e, phi) = 1
        BigInteger e = BigInteger.TWO;
        while (!gcd(e, phi).equals(BigInteger.ONE)) {
            e = e.add(BigInteger.ONE);
        }

        // Calculate d (private key) such that (d * e) % phi = 1
        BigInteger d = modInverse(e, phi);

        // Read plaintext from file
        BigInteger plaintext = input.nextBigInteger();

        // Encrypt: ciphertext = (plaintext^e) % n
        BigInteger ciphertext = modPow(plaintext, e, n);
        output.println("Ciphertext: " + ciphertext);

        // Decrypt: decrypted = (ciphertext^d) % n
        BigInteger decrypted = modPow(ciphertext, d, n);
        output.println("Decrypted text: " + decrypted);

        input.close();
        output.close();
    }
}

File: cp/c++/10_only_number.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
using namespace std;

// Function to calculate GCD
int gcd(int a, int b) {
    if (b == 0)
        return a;
    return gcd(b, a % b);
}

// Function to calculate (base^exp) % mod
long long modPow(long long base, long long exp, long long mod) {
    long long result = 1;
    while (exp > 0) {
        if (exp % 2 == 1)
            result = (result * base) % mod;
        base = (base * base) % mod;
        exp /= 2;
    }
    return result;
}

// Function to find modular inverse using Extended Euclidean Algorithm
long long modInverse(long long e, long long phi) {
    long long t = 0, newt = 1;
    long long r = phi, newr = e;
    while (newr != 0) {
        long long quotient = r / newr;
        t = t - quotient * newt;
        swap(t, newt);
        r = r - quotient * newr;
        swap(r, newr);
    }
    if (t < 0) t += phi;
    return t;
}

int main() {
    ifstream infile("10.txt");
    ofstream outfile("output10.txt");

    // Prime numbers (you should use larger primes in real RSA)
    long long p, q;
    infile >> p >> q;
    
    // Calculate n = p * q and phi = (p-1)*(q-1)
    long long n = p * q;
    long long phi = (p - 1) * (q - 1);

    // Choose e such that 1 < e < phi and gcd(e, phi) = 1
    long long e;
    for (e = 2; e < phi; e++) {
        if (gcd(e, phi) == 1)
            break;
    }

    // Calculate d (private key) such that (d * e) % phi = 1
    long long d = modInverse(e, phi);

    // Read plaintext from file
    long long plaintext;
    infile >> plaintext;

    // Encrypt: ciphertext = (plaintext^e) % n
    long long ciphertext = modPow(plaintext, e, n);
    outfile << "Ciphertext: " << ciphertext << endl;

    // Decrypt: decrypted = (ciphertext^d) % n
    long long decrypted = modPow(ciphertext, d, n);
    outfile << "Decrypted text: " << decrypted << endl;

    infile.close();
    outfile.close();
    return 0;
}
